### ISSUE TITLE
CommonJS module formatter: default export is assigned differently depending on whether there are any named exports

### ISSUE THREAD

**Issue by UltCombo** [babel/babel#2047]
This sample code:

```js
export default 1;
export const x = 2;
```

Generates:

```js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = 1;
var x = 2;
exports.x = x;
```

Which is fine.
Now if we remove the named export:

```js
export default 1;
```

Generates:

```js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = 1;
module.exports = exports["default"];
```

The `module.exports` is overwritten by the default export.
This still works fine if importing the module via `import` statements (due to the `interopRequireDefault` helper), but it is rather confusing and error-prone if importing the CJS module via `require()` calls.

That is, sometimes the default export is returned directly by `require('module')`, other times you will need to access the `require('module').default` property, all depending on an arbitrary factor that may change anytime (adding/removing named exports), which causes a breaking change for all consumers.

**Comment by loganfsmyth** [comment]
This is expected behavior: http://babeljs.io/docs/usage/modules/#interop

As documented there:

> If you don't want this behaviour then you can use the commonStrict module formatter.

**Comment by UltCombo** [comment]
Thanks for the quick reply.
I thought the strict module formatters were going to be removed (https://github.com/babel/babel/issues/587#issuecomment-71347804), I assume this has changed?

I'll be using the `commonStrict` formatter again now, thanks. :smile:

**Comment by UltCombo** [comment]
So, looks like the `commonStrict` module formatter does not attempt any interop at all -- I had forgot this bit.

In my ideal use case, I'd like to keep the ES/CJS modules interop, but have the CJS module formatter expose the default export in a consistent way.

Well, I guess for this rare use case (where I need dynamic `require()`s, can't use ES Modules syntax), I can just as well import and apply the `interop-require` helper from the babel-runtime package manually.

**Comment by danielo515** [comment]
I am also missing this feature. The reasoning behind removing it still unclear to me. Removing this feature almost breaks all the modules that trusted babel for transpile to regular CJS modules.

**Comment by loganfsmyth** [comment]
You can get this behavior in Babel 6 via https://www.npmjs.com/package/babel-plugin-add-module-exports

**Comment by danielo515** [comment]
Dear @loganfsmyth, I already found that plugin after a couple of hours of searching. So yes, why not add another plugin to the plethoria of plugins required for the most basic Babel functionality. I don't understand why you just give the option .

**Comment by xtuc** [comment]
@Daniel15 Would it help to indicate that on the website?

**Comment by danielo515** [comment]
Dear @xtuc, not helpful to me but definetively helpful for newcomers .

**Comment by UltCombo** [comment]
> I don't understand why you just give the option .

This must already have been explained several times in other places, but to put it in simple terms: 

Module `a.js`:

```js
export default { foo: 1 };
export const foo = 2;
```

Module `b.js`:

```js
console.log(require('./a').foo);
```

What do you expect this to print? The old behavior was inconsistent as the default export's properties sometimes could be referenced as named exports (when there are no named exports in the module), but the behavior would change as soon as a named export is added.

It is extremely dangerous that your existing imports would break when the exporter makes a completely unrelated change such as adding or removing named exports.

**Comment by danielo515** [comment]
So it is better to break in a backwards manner (breaking all the modules that relies on this functionality) than in forwards manner, even taking into account that some thousands will never export a second property.

In any case, what you said makes a lot of sense and it is logically argued, so I can not be against it. Making breaking changes to change things that makes non sense should happen more often, I say it several times, it is ridiculous that now I'm complaining about one of those.

Thanks and regards

**Comment by loganfsmyth** [comment]
We can definitely do better on documentation on this, sorry about that.

The point mentioned above is indeed the problem we looked to solve. I've also elaborated a bunch on that here if you are curious: http://stackoverflow.com/questions/33505992/babel-6-changes-how-it-exports-default

We made this change on a Babel major version so that it would only affect people as they upgrade, and since it was a behavior that was easy to add back via a standalone plugin, we felt it was a good change to make. 

This behavior will also likely become more common moving forward when ES6 module support is added to Node itself.
